src/java/org/apache/cassandra/index/sai/utils/SegmentMemoryLimiter.java

```diff
@@ -0,0 +1,122 @@
+     * <p>
+     * Note that to avoid flushing small index segments, a segment is only flushed when
+     * both the global size of all building segments has breached the limit and the size of the
+     * segment in question reaches (segment_write_buffer_space_mb / # currently building column indexes).
+     * <p>
+     * ex. If there is only one column index building, it can buffer up to segment_write_buffer_space_mb.
+     * <p>
+     * ex. If there is one column index building per table across 8 compactors, each index will be
+     *     eligible to flush once it reaches (segment_write_buffer_space_mb / 8) MBs.
+     */
+    public static final long DEFAULT_SEGMENT_BUILD_MEMORY_LIMIT = DatabaseDescriptor.getSAISegmentWriteBufferSpace().toBytes();
+
+    private static final AtomicLong bytesUsed = new AtomicLong(0);
+
+    private static volatile long limitBytes = DEFAULT_SEGMENT_BUILD_MEMORY_LIMIT;
+
+    private static volatile long minimumFlushSize = 0;
+
+    /** The number of column indexes being built globally. (Starts at one to avoid divide by zero.) */
+    private static final AtomicInteger activeBuilderCount = new AtomicInteger(0);
+
+    public static int getActiveBuilderCount()
+    {
+        return activeBuilderCount.get();
+    }
+
+    public static void registerBuilder()
+    {
+        minimumFlushSize =  limitBytes / activeBuilderCount.incrementAndGet();
+    }
+
+    public static void unregisterBuilder()
+    {
+        int builders = activeBuilderCount.decrementAndGet();
+        minimumFlushSize = builders == 0 ? limitBytes : limitBytes / builders;
+    }
+
+    /**
+     * @return true if the current number of bytes allocated against the tracker has breached the limit, false otherwise
+     */
+    public static boolean usageExceedsLimit(long bytesUsed)
+    {
+        return currentBytesUsed() > limitBytes && bytesUsed > minimumFlushSize;
+    }
+
+    public static long increment(long bytes)
+    {
+        if (logger.isTraceEnabled())
+            logger.trace("Incrementing tracked memory usage by {} bytes from current usage of {}...", bytes, currentBytesUsed());
+        return bytesUsed.addAndGet(bytes);
+    }
+
+    public static long decrement(long bytes)
+    {
+        if (logger.isTraceEnabled())
+            logger.trace("Decrementing tracked memory usage by {} bytes from current usage of {}...", bytes, currentBytesUsed());
+        return bytesUsed.addAndGet(-bytes);
+    }
+
+    public static long currentBytesUsed()
+    {
+        return bytesUsed.get();
+    }
+    
+    public static long limitBytes()
+    {
+        return limitBytes;
+    }
+
+    @VisibleForTesting
+    public static void setLimitBytes(long bytes)
+    {
+        limitBytes = bytes;
+    }
+
+    @VisibleForTesting
+    public static void reset()
+    {
+        bytesUsed.set(0);
+        limitBytes = DEFAULT_SEGMENT_BUILD_MEMORY_LIMIT;
+        activeBuilderCount.set(0);
```

Comment:
I obviously would rather move away from having everything here be static, but if we do leave it that way, for completeness, I guess adding `minimumFlushSize = 0;` here would be a good idea...

(see https://github.com/apache/cassandra/pull/2498/files#r1270102370)

Reply:
I have added this, but it no longer exists in `SegmentMemoryLimiter`. This is now a singleton and is overridden by `SAITester$TestMemoryLimiter` where this method now lives.